[Clang][Driver] Revise Cygwin ToolChain to call linker directly

tyan0 opened the pull request:
...so that `libc++`, `compiler-rt`, and `libunwind` can be used by the options: `-stdlib=libc++`, `-rtlib=compiler-rt`, and `-unwindlib=libunwind` respectively. Along with this change, the test for this driver is also trimmed a bit.

This is a followup patch for https://github.com/llvm/llvm-project/commit/52924a2d7255cdd280b2b82dad8616e01fe065da.

tyan0 committed:
Revise Cygwin ToolChain to call linker directly

...so that libc++, compiler-rt, and libunwind can be used by the
options: -stdlib=libc++, -rtlib=compiler-rt, and -unwindlib=libunwind
respectively. Along with this change, the test for this driver is also
trimmed a bit.

This is a follow-up patch to the commit 52924a2d7255.

Signed-off-by: Takashi Yano <takashi.yano@nifty.ne.jp>

jeremyd2019 commented:
Thanks!  This was on my mental TODO list but I haven't gotten back to looking at it.  When I was thinking about this, I was thinking about modifying the MinGW linker with a couple of virtual functions and deriving from it for a Cygwin linker, to avoid so much code duplication.  Maybe @mstorsjo can provide an opinion here?

Also, the cygwin.cpp test was based on hurd.cpp, I'm thinking more of those tests are relevant now.

kikairoya commented on clang/lib/Driver/ToolChains/Cygwin.cpp:
```diff
@@ -107,3 +110,289 @@ void Cygwin::AddClangSystemIncludeArgs(const ArgList &DriverArgs,
+
+    if (IsShared) {
+      CmdArgs.push_back("-e");
+      CmdArgs.push_back("_cygwin_dll_entry");
```
for x86, _cygwin_dll_entry@12

jeremyd2019 commented on clang/lib/Driver/ToolChains/Cygwin.cpp:
```diff
@@ -107,3 +110,289 @@ void Cygwin::AddClangSystemIncludeArgs(const ArgList &DriverArgs,
+
+  ToolChain.AddFilePathLibArgs(Args, CmdArgs);
+
+  CmdArgs.push_back(Args.MakeArgString(
+      "-L" + ToolChain.getGCCInstallation().getInstallPath()));
```
I don't think you should need this (or the static_cast of the ToolChain), adding the install path to the FIlePaths in the Cygwin constructor should have been sufficient.

tyan0 replied:
Thanks! Removed as well as `getGCCInstallation()` method.

kikairoya commented:
Could you let I know output of `echo | clang -xc - -v` (and some option variants) with patch?

Does this patch consider to use of LLD? ( gcc can use with LLD: `-fuse-ld=lld -L/usr/lib/w32api -Wl,--enable-auto-import,--disable-high-entropy-va,--disable-nxcompat,--wrap,_Znwm,--wrap,_Znam,--wrap,_ZdlPv,--wrap,_ZdaPv,--wrap,_ZnwmRKSt9nothrow_t,--wrap,_ZnamRKSt9nothrow_t,--wrap,_ZdlPvRKSt9nothrow_t,--wrap,_ZdaPvRKSt9nothrow_t,-u,_Znwm,-u,_Znam,-u,_ZdlPv,-u,_ZdaPv,-u,_ZnwmRKSt9nothrow_t,-u,_ZnamRKSt9nothrow_t,-u,_ZdlPvRKSt9nothrow_t,-u,_ZdaPvRKSt9nothrow_t` )

> I was planning to refer to that, MinGW.cpp, and the GCC spec for Cygwin, and see if I could come up with a couple of places where virtual functions could be added to the MinGW::Linker class so that Cygwin::Linker could derive from it and override (I think the "wrap" stuff and list of default libraries)

+1
I think Cygwin-specific driver will not be maintained by others and will diverge and left alone from other platform or components.
But, unfortunately, differences of MinGW and Cygwin are scattered widely so doing that isn't trivial I think.
My hack (copy from MinGW.cpp) may help to see difference between MinGW and Cygwin. https://github.com/kikairoya/llvm-project/blob/cygwin-new-driver/clang/lib/Driver/ToolChains/Cygwin.cpp

jeremyd2019 commented:
> > I was planning to refer to that, MinGW.cpp, and the GCC spec for Cygwin, and see if I could come up with a couple of places where virtual functions could be added to the MinGW::Linker class so that Cygwin::Linker could derive from it and override (I think the "wrap" stuff and list of default libraries)
> 
> +1 I think Cygwin-specific driver will not be maintained by others and will diverge and left alone from other platform or components. But, unfortunately, differences of MinGW and Cygwin are scattered widely so doing that isn't trivial I think. My hack (copy from MinGW.cpp) may help to see difference between MinGW and Cygwin. https://github.com/kikairoya/llvm-project/blob/cygwin-new-driver/clang/lib/Driver/ToolChains/Cygwin.cpp

I meant just the Linker class.  I don't think it's really possible to avoid a dedicated Cygwin driver, but I'm hoping it can keep close to the Generic_GCC base class rather than redoing so much stuff as MinGW did.

tyan0 commented on clang/lib/Driver/ToolChain.cpp:
```diff
@@ -1504,6 +1506,7 @@ void ToolChain::AddCXXStdlibLibArgs(const ArgList &Args,
   switch (Type) {
   case ToolChain::CST_Libcxx:
     CmdArgs.push_back("-lc++");
+    CmdArgs.push_back("-lc++abi");
```
I'm not sure. `libc++.dll.a` depends `libc++abi.dll.a`. Similarly, `libc++.so.1` depends `libc++abi.so.1` in linux. Why linux does not need `-lc++abi` here???
```
$ ldd /usr/lib/llvm-14/lib/libc++.so.1
        linux-vdso.so.1 (0x00007fffe5f04000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f203b74e000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f203b66e000)
        libc++abi.so.1 => /lib/x86_64-linux-gnu/libc++abi.so.1 (0x00007f203b636000)
        libunwind.so.1 => /lib/x86_64-linux-gnu/libunwind.so.1 (0x00007f203b62b000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f203ba46000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f203b60b000)
```

kikairoya replied:
ELF DSO can embed its dependencies in itself and find them at link time. PE/COFF doesn't have such mechanism for the link time AFAIK.

[MinGW package is built with -DLIBCXX_ENABLE_STATIC_ABI_LIBRARY](https://github.com/mstorsjo/llvm-mingw/blob/a1f6413e5c21fd74b64137b56167f4fba500d1d8/build-libcxx.sh#L99) so doesn't need `-lc++abi`.

kikairoya replied:
It's packaging policy matter, but it would be better to merge libc++abi in libc++ because libc++ can't link to another ABI library. MinGW does so.

kikairoya commented:
Thanks.

It looks like some linker options might still be missing or need adjustments:

- `--enable-auto-import`, `--disable-high-entropy-va`, and `--dll-search-prefix=cyg` are generally required on Cygwin.
- `-pie` has no effect when targeting PE/COFF and can be ignored.

Also, a few options should be handled in both `class Cygwin` and in the linker job construction:

- `-mwindows` should be translated into `--subsystem=windows`.
- `-shared` and `-mdll` are subtly different in what they instruct GCC to link with.

> I confirmed 'clang++ -fuse-ld=lld` works. What is your concern about LLD?

Here are a few issues I've seen when using LLD:

- LLD requires passing `-u _Znwm` and other operators explicitly, otherwise [`default_cygwin_cxx_malloc` defined in `libcygwin.a`](https://github.com/mirror/newlib-cygwin/blob/1b7c72fdcc4bde7520407d2d3364146f04fb8312/winsup/cygwin/cxx.cc#L92) pulls wrong reference (at least as of a month ago), due to [different handling of `--wrap` compared to GNU ld](https://maskray.me/blog/2020-12-19-lld-and-gnu-linker-incompatibilities). This may surface in `check-cxx` failures.
- LLD doesn’t have built-in default search paths, so `-L/usr/lib/w32api` needs to be specified.
- [Binaries linked with LLD can't be rebased](https://cygwin.com/pipermail/cygwin/2025-June/258268.html), which means we shouldn’t use `--disable-dynamicbase` (currently it's fine, as this option isn't set).

It would be great if this patch could be adjusted to support bootstrapping without requiring extra linker flags.

kikairoya commented:
If we should consider cross-compiling, it's needed to be aware that `$SYSROOT/usr/lib` doesn't appear in real filesystem.

jeremyd2019 commented on clang/lib/Driver/ToolChains/Cygwin.cpp:
```diff
@@ -107,3 +110,282 @@ void Cygwin::AddClangSystemIncludeArgs(const ArgList &DriverArgs,
+
+    if (IsShared) {
+      CmdArgs.push_back("-e");
+      CmdArgs.push_back(ToolChain.getTriple().isArch32Bit()
```
This should only be for x86, not all 32-bit arches (though that isn't really important for Cygwin, which isn't expected to handle any additional 32-bit arches)
```suggestion
      CmdArgs.push_back(ToolChain.getArch() == llvm::Triple::x86
```
Based on what MinGW.cpp does

tyan0 replied:
I see. Updated.

tyan0 commented:
> It looks like some linker options might still be missing or need adjustments:
> 
> * `--enable-auto-import`, `--disable-high-entropy-va`, and `--dll-search-prefix=cyg` are generally required on Cygwin.
> * `-pie` has no effect when targeting PE/COFF and can be ignored.
> 
> Also, a few options should be handled in both `class Cygwin` and in the linker job construction:
> 
> * `-mwindows` should be translated into `--subsystem=windows`.
> * `-shared` and `-mdll` are subtly different in what they instruct GCC to link with.

I have been convinced by your points regarding `-dll-search-prefix=cyg`, `-pie`, `-mwindows`, `-shared`, and `-mdll`. These have been implemented to the PR.

However, as for `--enable-auto-import` and `--disable-high-entropy-va`, gcc/g++ do not add these flags by default. I doubt these are really necessary. Why do you think these are needed?

jeremyd2019 commented:
high-entropy-va would indeed break fork.  I don't know about nxcompat.

Here is a patch I had to lld to try to change defaults to match Cygwin defaults in BFD, but obviously it doesn't work out because lld can't tell if it's cygwin or mingw.
31a9812c6ad9e69f840506b65fcf04085cac7cdf

I don't know if you want to try to handle all of these in the cygwin driver.  It'd probably be the best place to do it since it does know that it is Cygwin and not MinGW.  But it would mean that using gcc -fuse-ld=lld wouldn't work due to wrong defaults.

tyan0 commented:
> > However, as for `--enable-auto-import` and `--disable-high-entropy-va`, gcc/g++ do not add these flags by default. I doubt these are really necessary. Why do you think these are needed?
> 
> I was confused and not accurate earlier, sorry about that.
> 
> Former, `--enable-auto-import` is not needed by default but need to handle for consistency with MinGW.
> 
> Latter, `--disable-high-entropy-va`, is needed by LLD since BFD-ld disable it by default. `--disable-nxcompat` may also be needed, specify this would be better. High entropy VA ( or, NX bit, I haven't dug it ) causes `fork` to fail.

I could not reproduce fork() failure even with high-entropy-va.
```
$ cat hello.cc
#include <unistd.h>
#include <iostream>

int main() {
  fork();
  std::cout << "Hello" << std::endl;
  return 0;
}
$ clang++ -fuse-ld=lld hello.cc
$ peflags -v a.exe
a.exe: coff(0x0022[+executable_image,+bigaddr]) pe(0x8160[+high-entropy-va,+dynamicbase,+nxcompat,+tsaware])
a.exe: Checksum update failed
$ ./a.exe
Hello
Hello
$
```

tyan0 commented:
> high-entropy-va would indeed break fork. I don't know about nxcompat.

Added `--disable-high-entropy-va` when neither `--disable-high-entropy-va` nor `--high-entropy-va` is specified.

jeremyd2019 commented:
This will need more tests added to cover these additional flags that are now hooked up.  Off the top of my head, `-shared`/`-mdll` and the `_cygwin_dll_entry` x86_64 vs i686, `-mconsole`/`-mwindows`, the high-entropy-va thing.  Might also be worthwhile to look at the hurd.cpp driver test and see if any more of that is relevant now, that's what I started with to make the cygwin.cpp test.

Is the `_cygwin_dll_entry` option being added for `-mdll` or only for `-shared`?  It looks like gcc does it for both, but only changes to `crtbeginS.o` for `-shared`

kikairoya commented:
> $ clang++ -fuse-ld=lld hello.cc

Do you have patched to LLD locally? This should fail with `unable to find library -lkernel32`.


> But it would mean that using gcc -fuse-ld=lld wouldn't work due to wrong defaults.

LLD and Clang are both part of LLVM, so they should be expected to work together without requiring extra hacky options. For GCC, we can patch its driver downstream if necessary -- though ideally, this should be addressed upstream, assuming the GCC maintainers are open to it.

kikairoya commented on clang/lib/Driver/ToolChains/Cygwin.cpp:
```diff
@@ -107,3 +110,294 @@ void Cygwin::AddClangSystemIncludeArgs(const ArgList &DriverArgs,
+  const Driver &D = ToolChain.getDriver();
+
+  const bool IsIAMCU = ToolChain.getTriple().isOSIAMCU();
+  const bool IsVE = ToolChain.getTriple().isVE();
```
VE means https://www.phoronix.com/news/LLVM-NEC-VE-Official ? Cygwin can't run on it.

tyan0 replied:
Removed. Thanks.

kikairoya commented on clang/lib/Driver/ToolChains/Cygwin.cpp:
```diff
@@ -107,3 +110,294 @@ void Cygwin::AddClangSystemIncludeArgs(const ArgList &DriverArgs,
+  const auto &ToolChain = getToolChain();
+  const Driver &D = ToolChain.getDriver();
+
+  const bool IsIAMCU = ToolChain.getTriple().isOSIAMCU();
```
Cygwin is not a IAMCU OS.

tyan0 replied:
Removed. Thanks.

kikairoya commented on clang/lib/Driver/ToolChains/Cygwin.cpp:
```diff
@@ -107,3 +110,294 @@ void Cygwin::AddClangSystemIncludeArgs(const ArgList &DriverArgs,
+
+static bool getStatic(const ArgList &Args) {
+  return Args.hasArg(options::OPT_static) &&
+         !Args.hasArg(options::OPT_static_pie);
```
static-pie has no effect.

tyan0 replied:
Removed. Thanks.

kikairoya commented on clang/lib/Driver/ToolChains/Cygwin.cpp:
```diff
@@ -107,3 +110,294 @@ void Cygwin::AddClangSystemIncludeArgs(const ArgList &DriverArgs,
+    CmdArgs.push_back("thumb2pe");
+    break;
+  case llvm::Triple::aarch64:
+    if (ToolChain.getEffectiveTriple().isWindowsArm64EC())
```
It would be better to list only targets that are supported by Cygwin or reflect the MinGW list.

MinGW now support a new arch:
https://github.com/llvm/llvm-project/blob/a102342990231f8558361da68e5df92c7b1c737d/clang/lib/Driver/ToolChains/MinGW.cpp#L134-L136

tyan0 replied:
Done. Thanks.

jeremyd2019 commented:
> > $ clang++ -fuse-ld=lld hello.cc
> 
> Do you have patched to LLD locally? This should fail with `unable to find library -lkernel32`.

this new driver should be passing all the necessary `-L` paths to the linker, including `/usr/lib/w32api`

tyan0 commented on clang/lib/Driver/ToolChains/Cygwin.cpp:
```diff
@@ -107,3 +110,294 @@ void Cygwin::AddClangSystemIncludeArgs(const ArgList &DriverArgs,
+
+  const bool IsIAMCU = ToolChain.getTriple().isOSIAMCU();
+  const bool IsVE = ToolChain.getTriple().isVE();
+  const bool IsStaticPIE = getStaticPIE(Args, ToolChain);
```
Removed. Thanks.

tyan0 commented:
> This will need more tests added to cover these additional flags that are now hooked up. Off the top of my head, `-shared`/`-mdll` and the `_cygwin_dll_entry` x86_64 vs i686, `-mconsole`/`-mwindows`, the high-entropy-va thing. Might also be worthwhile to look at the hurd.cpp driver test and see if any more of that is relevant now, that's what I started with to make the cygwin.cpp test.
> 
> Is the `_cygwin_dll_entry` option being added for `-mdll` or only for `-shared`? It looks like gcc does it for both, but only changes to `crtbeginS.o` for `-shared`

Thanks. Test updated.

jeremyd2019 commented:
I noticed another bit missing from the MinGW driver: https://github.com/llvm/llvm-project/blob/f761d73265119eeb3b1ab64543e6d3012078ad13/clang/lib/Driver/ToolChains/MinGW.cpp#L198-L208

kikairoya commented on clang/lib/Driver/ToolChains/Cygwin.cpp:
```diff
@@ -107,3 +110,294 @@ void Cygwin::AddClangSystemIncludeArgs(const ArgList &DriverArgs,
+
+  bool saw_high_entropy_va = false;
+  for (const char *Arg : CmdArgs)
+    if (StringRef(Arg) == "--high-entropy-va" ||
```
Since the linkers adopt last-flag-wins rule, simply calling `CmdArgs.push_back("--disable-high-entropy-va")` before calling `AddLinkerInputs` is enough to work. `-Wl,` is not needed to be handled explicitly.

tyan0 commented:
> I noticed another bit missing from the MinGW driver:
> 
> https://github.com/llvm/llvm-project/blob/f761d73265119eeb3b1ab64543e6d3012078ad13/clang/lib/Driver/ToolChains/MinGW.cpp#L198-L208

Thanks. Applied with additional test.

kikairoya commented:
> * LLD requires passing `-u _Znwm` and other operators explicitly, otherwise [`default_cygwin_cxx_malloc` defined in `libcygwin.a`](https://github.com/mirror/newlib-cygwin/blob/1b7c72fdcc4bde7520407d2d3364146f04fb8312/winsup/cygwin/cxx.cc#L92) pulls wrong reference (at least as of a month ago), due to [different handling of `--wrap` compared to GNU ld](https://maskray.me/blog/2020-12-19-lld-and-gnu-linker-incompatibilities). This may surface in `check-cxx` failures.

This also happened today, and not only with LLD but also with BFD, so "due to different handling of `--wrap` compared to GNU ld" seems wrong. Unfortunately, I couldn't create a minimal reproducer, so I can't tell whether the issue lies in the linkers, libc++, or libcygwin.a.
Other functionalities look good to me; this patch enables bootstrapping using LLD without hacky linker flags.

tyan0 commented:
> > * LLD requires passing `-u _Znwm` and other operators explicitly, otherwise [`default_cygwin_cxx_malloc` defined in `libcygwin.a`](https://github.com/mirror/newlib-cygwin/blob/1b7c72fdcc4bde7520407d2d3364146f04fb8312/winsup/cygwin/cxx.cc#L92) pulls wrong reference (at least as of a month ago), due to [different handling of `--wrap` compared to GNU ld](https://maskray.me/blog/2020-12-19-lld-and-gnu-linker-incompatibilities). This may surface in `check-cxx` failures.
> 
> This also happened today, and not only with LLD but also with BFD, so "due to different handling of `--wrap` compared to GNU ld" seems wrong. Unfortunately, I couldn't create a minimal reproducer, so I can't tell whether the issue lies in the linkers, libc++, or libcygwin.a. Other functionalities look good to me; this patch enables bootstrapping using LLD without hacky linker flags.

I am not sure why the issue happens; however, I doubt whether `-u _Znwm` etc. make any real difference.
```man
       -u symbol
       --undefined=symbol
           Force  symbol  to  be  entered  in  the  output file as an undefined
           symbol.  Doing this may, for example, trigger linking of  additional
           modules  from standard libraries.  -u may be repeated with different
           option arguments to enter additional undefined symbols.  This option
           is equivalent to the "EXTERN" linker script command.

           If this option is being used  to  force  additional  modules  to  be
           pulled into the link, and if it is an error for the symbol to remain
           undefined, then the option --require-defined should be used instead.
```
The symbol `_Znwm` itself does not exist in any libraries or modules, I think.

tyan0 commented:
> I mean many parts of `check-cxx` crash. `cygc++-1.dll` indeed exports `_Znwm`.

I see. _Znwm etc. are mangled name for `new` and `delete`.
```
_Znwm: operator new
_Znam: operator new[]
_ZdlPv: operator delete
_ZdaPv: operator delete[]
_ZnwmRKSt9nothrow_t: operator new(std::nothrow)
_ZnamRKSt9nothrow_t: operator new[](std::nothrow)
_ZdlPvRKSt9nothrow_t: operator delete (for new(nothrow))
_ZdaPvRKSt9nothrow_t: operator delete[] (for new[](nothrow))
```
Therefore, `liststdc++.a` and `libc++.a` have these symbols.

So, what is the meaning of linking these symbols explicitly?

jeremyd2019 commented:
I was thinking that `-u` wouldn't do anything useful (and I'm still thinking that), but I just had a thought about why you might think it's necessary (or, why it might not work as expected).

`--wrap` renames the definition to `__real_X` and references to `__wrap_X`.  If libc++ has the operator new/delete declared as dllimport, the symbol involved would actually be `__imp_X`, which wouldn't be redirected by `--wrap`.  `-u` might be trying to force the inclusion of the jump stub, which would be wrapped.  But, `__imp_X` is still unwrapped and might be used directly.  Does this make sense?

jeremyd2019 commented:
I made a quick test and I'm seeing `_ZdlPvm` being used for `operator delete` and that is not being wrapped.  That may need to be wrapped too, because it calls `__wrap__Znwm` which calls `default_cygwin_cxx_malloc`, but `_ZdlPvm` which is the version from libc++.dll.

jeremyd2019 commented:
Nope, that's not the problem.  The problem is that this relies on weak symbols, and that they end up as `NULL` if not explicitly defined.

Test code:
```C++
#include <stdio.h>
int main (void)
{
        char *p = new char (0);
        printf ("%p\n", p);
        delete p;
        return 0;
}
```

With ld.bfd:
```
(gdb) p __cygwin_cxx_malloc
$2 = {oper_new = 0x0, oper_new__ = 0x0, oper_delete = 0x0,
  oper_delete__ = 0x0, oper_new_nt = 0x0, oper_new___nt = 0x0,
  oper_delete_nt = 0x0, oper_delete___nt = 0x0}
```

with ld.lld:
```
(gdb) p __cygwin_cxx_malloc
$1 = {oper_new = 0x7ff7f41710c8 <operator new(unsigned long)>,
  oper_new__ = 0x7ff6b4170000, oper_delete = 0x7ff6b4170000,
  oper_delete__ = 0x7ff6b4170000, oper_new_nt = 0x7ff6b4170000,
  oper_new___nt = 0x7ff6b4170000, oper_delete_nt = 0x7ff6b4170000,
  oper_delete___nt = 0x7ff6b4170000}
(gdb) p *__cygwin_cxx_malloc.oper_new_nt
Cannot access memory at address 0x7ff6b4170000
```

Notably, those ending in `0000` are invalid memory.

Weak symbols do not work out well in PE/COFF, but apparently Cygwin is relying on them working for this.

mstorsjo commented:
> Weak symbols do not work out well in PE/COFF

They work just fine in clang+lld environments for mingw targets, but there's a handful of known issues relating to them with gcc+ld.bfd, for mingw. But apparently those issues don't hit cygwin targets.

jeremyd2019 commented:
> > Do the same for dynamicbase and nxcompat as a precaution.
> 
> dynamicbase is needed for LLD.

only because --enable-auto-image-base and the differing default image bases for Cygwin are not implemented in LLD

jeremyd2019 commented:
> > That said, I also encountered the same issue with BFD-ld (when building libc++ with @tyan0's patches from cygwin-packages), so I should try to reproduce it again to be sure. It may take a few days to rebuild and investigate...
> 
> What I saw is probably a bug in LLD, and not related to this PR anyway.

https://cygwin.com/pipermail/cygwin/2025-July/258547.html
Probably need to open a new issue for the LLD weirdness.

tyan0 commented:
> > > Do the same for dynamicbase and nxcompat as a precaution.
> > 
> > 
> > dynamicbase is needed for LLD.
> 
> only because --enable-auto-image-base and the differing default image bases for Cygwin are not implemented in LLD

Thanks for explanation. I'll drop the code to let `--disable-dynamicbase` default.

tyan0 commented on clang/lib/Driver/ToolChains/Cygwin.cpp:
```diff
@@ -107,3 +110,282 @@ void Cygwin::AddClangSystemIncludeArgs(const ArgList &DriverArgs,
+    CmdArgs.push_back("console");
+  }
+
+  CmdArgs.push_back("--wrap=_Znwm");
+  CmdArgs.push_back("--wrap=_Znam");
+  CmdArgs.push_back("--wrap=_ZdlPv");
+  CmdArgs.push_back("--wrap=_ZdaPv");
+  CmdArgs.push_back("--wrap=_ZnwmRKSt9nothrow_t");
+  CmdArgs.push_back("--wrap=_ZnamRKSt9nothrow_t");
+  CmdArgs.push_back("--wrap=_ZdlPvRKSt9nothrow_t");
+  CmdArgs.push_back("--wrap=_ZdaPvRKSt9nothrow_t");
```
Thanks. Is this true for all 32bit platform? I mean, also for 32bit arm? Or only for i686?

tyan0 replied:
I see. This should be for all 32bit platforms.

tyan0 replied:
_Znwy? not _Znwn? For unsigned long long, _Znwn is used IIUC. _Znwy is for unsigned short, isn't it?

jeremyd2019 replied:
that is right for aarch64-w64-mingw32 but a cygwin target should override that because it is LP64 not LLP64 like Windows/mingw

jeremyd2019 replied:
This may be a silly idea, but clang obviously knows how to mangle names - is this API accessible to the driver?

tyan0 replied:
Like this?
https://clang.llvm.org/doxygen/classclang_1_1MangleContext.html#a74417e8b66965c5e167bd5faf21df9a1

kikairoya replied:
`GlobalDecl` isn't accessible from this context. The `Driver` component can reference only the `Basic` component.

However, I found a way to see what `size_t` does mean on target triple (see diff).
<details>

<summary>diff</summary>

```diff
diff --git a/clang/lib/Driver/ToolChains/Cygwin.cpp b/clang/lib/Driver/ToolChains/Cygwin.cpp
index 079911b0f0ac..e2561172d386 100644
--- a/clang/lib/Driver/ToolChains/Cygwin.cpp
+++ b/clang/lib/Driver/ToolChains/Cygwin.cpp
@@ -7,6 +7,7 @@
 //===----------------------------------------------------------------------===//

 #include "Cygwin.h"
+#include "clang/Basic/TargetInfo.h"
 #include "clang/Config/config.h"
 #include "clang/Driver/CommonArgs.h"
 #include "clang/Driver/Compilation.h"
@@ -177,24 +178,32 @@ void cygwin::Linker::ConstructJob(Compilation &C, const JobAction &JA,
     CmdArgs.push_back("console");
   }

-  if (ToolChain.getTriple().isArch32Bit()) {
-    CmdArgs.push_back("--wrap=_Znwj");
-    CmdArgs.push_back("--wrap=_Znaj");
-    CmdArgs.push_back("--wrap=_ZdlPv");
-    CmdArgs.push_back("--wrap=_ZdaPv");
-    CmdArgs.push_back("--wrap=_ZnwjRKSt9nothrow_t");
-    CmdArgs.push_back("--wrap=_ZnajRKSt9nothrow_t");
-    CmdArgs.push_back("--wrap=_ZdlPvRKSt9nothrow_t");
-    CmdArgs.push_back("--wrap=_ZdaPvRKSt9nothrow_t");
-  } else {
-    CmdArgs.push_back("--wrap=_Znwm");
-    CmdArgs.push_back("--wrap=_Znam");
-    CmdArgs.push_back("--wrap=_ZdlPv");
-    CmdArgs.push_back("--wrap=_ZdaPv");
-    CmdArgs.push_back("--wrap=_ZnwmRKSt9nothrow_t");
-    CmdArgs.push_back("--wrap=_ZnamRKSt9nothrow_t");
-    CmdArgs.push_back("--wrap=_ZdlPvRKSt9nothrow_t");
-    CmdArgs.push_back("--wrap=_ZdaPvRKSt9nothrow_t");
+  auto TI = targets::AllocateTarget(ToolChain.getTriple(), {});
+  const auto TypeChar = [&]() {
+    switch (TI->getSizeType()) {
+    case TransferrableTargetInfo::UnsignedShort:
+      return 't';
+    case TransferrableTargetInfo::UnsignedInt:
+      return 'j';
+    case TransferrableTargetInfo::UnsignedLong:
+      return 'm';
+    case TransferrableTargetInfo::UnsignedLongLong:
+      return 'y';
+    default:
+      llvm_unreachable("Invalid SizeType");
+    }
+  }();
+  const auto *const LabelPrefix = TI->getUserLabelPrefix();
+  std::string WrapPrefix = "--wrap=";
+  if (LabelPrefix)
+    WrapPrefix += LabelPrefix;
+  for (std::string S : {"_Znw%", "_Zna%", "_ZdlPv", "_ZdaPv",
+                        "_Znw%RKSt9nothrow_t", "_Zna%RKSt9nothrow_t",
+                        "_ZdlPvRKSt9nothrow_t", "_ZdaPvRKSt9nothrow_t"}) {
+    auto Pos = S.find('%');
+    if (Pos != S.npos)
+      S[Pos] = TypeChar;
+    CmdArgs.push_back(Args.MakeArgString(WrapPrefix + S));
   }

   if (Args.hasArg(options::OPT_mdll))
```
</details>

BTW, `gcc -m32` doesn't wrap `new` properly (perhaps the old i686-native gcc can?) and the different mangling rule between `new` and `delete` is found from `libstdxx_wrapper.cc` in `cygwin1.dll`.

```
$ echo 'int main() { auto p = new int(3); delete p; }' | g++ -m64 -xc++ - -nostartfiles -nodefaultlibs -Wl,--no-demangle,--entry, -shared -fno-sized-deallocation
/usr/lib/gcc/x86_64-pc-cygwin/16/../../../../x86_64-pc-cygwin/bin/ld: /tmp/ccc30TZw.o:<stdin>:(.text+0x9): undefined reference to `__main'
/usr/lib/gcc/x86_64-pc-cygwin/16/../../../../x86_64-pc-cygwin/bin/ld: /tmp/ccc30TZw.o:<stdin>:(.text+0x13): undefined reference to `__wrap__Znwm'
/usr/lib/gcc/x86_64-pc-cygwin/16/../../../../x86_64-pc-cygwin/bin/ld: /tmp/ccc30TZw.o:<stdin>:(.text+0x2e): undefined reference to `__wrap__ZdlPv'
/usr/lib/gcc/x86_64-pc-cygwin/16/../../../../x86_64-pc-cygwin/bin/ld: /tmp/ccc30TZw.o:<stdin>:(.text+0x3f): undefined reference to `__wrap__ZdlPv'
collect2: エラー: ld はステータス 1 で終了しました


$ echo 'int main() { auto p = new int(3); delete p; }' | g++ -m32 -xc++ - -nostartfiles -nodefaultlibs -Wl,--no-demangle,--entry, -shared -fno-sized-deallocation
/usr/lib/gcc/x86_64-pc-cygwin/16/../../../../x86_64-pc-cygwin/bin/ld: /tmp/ccAA6tBm.o:<stdin>:(.text+0x7): undefined reference to `___main'
/usr/lib/gcc/x86_64-pc-cygwin/16/../../../../x86_64-pc-cygwin/bin/ld: /tmp/ccAA6tBm.o:<stdin>:(.text+0x13): undefined reference to `__Znwj'
/usr/lib/gcc/x86_64-pc-cygwin/16/../../../../x86_64-pc-cygwin/bin/ld: /tmp/ccAA6tBm.o:<stdin>:(.text+0x2d): undefined reference to `___wrap__ZdlPv'
/usr/lib/gcc/x86_64-pc-cygwin/16/../../../../x86_64-pc-cygwin/bin/ld: /tmp/ccAA6tBm.o:<stdin>:(.text+0x3c): undefined reference to `___wrap__ZdlPv'
collect2: エラー: ld はステータス 1 で終了しました
```

jeremyd2019 replied:
That's interesting.  I like that in the context of having to add more overloads eventually (once Cygwin and GCC catch up):
```diff
++    CmdArgs.push_back("--wrap=_ZdaPvm");
++    CmdArgs.push_back("--wrap=_ZdlPvm");
++    CmdArgs.push_back("--wrap=_ZnwmSt11align_val_t");
++    CmdArgs.push_back("--wrap=_ZnamSt11align_val_t");
++    CmdArgs.push_back("--wrap=_ZdlPvSt11align_val_t");
++    CmdArgs.push_back("--wrap=_ZdlPvmSt11align_val_t");
++    CmdArgs.push_back("--wrap=_ZdaPvSt11align_val_t");
++    CmdArgs.push_back("--wrap=_ZdaPvmSt11align_val_t");
++    CmdArgs.push_back("--wrap=_ZnwmSt11align_val_tRKSt9nothrow_t");
++    CmdArgs.push_back("--wrap=_ZnamSt11align_val_tRKSt9nothrow_t");
++    CmdArgs.push_back("--wrap=_ZdlPvSt11align_val_tRKSt9nothrow_t");
++    CmdArgs.push_back("--wrap=_ZdaPvSt11align_val_tRKSt9nothrow_t");
```

kikairoya replied:
> `gcc -m32` doesn't wrap `new` properly

I found that gcc tries to wrap `operator new(unsigned long)` even if `-m32`.


> Oh, also it doesn't seem that you need to prefix `_` on these `--wrap` arguments on i686.

Ah, true. `_` will be added by the linker. This mangling confuses me so much...

jeremyd2019 commented:
There seems to be some other issue with this patch on 32-bit that I need to track down: when building llvm/clang/compiler-rt 21 with this patch with a clang 21 with this patch, I get errors building compiler-rt builtins (which uses the freshly-built clang).  The fact that this only happens on the second build is confusing, because it does use the freshly-built clang in this case.  I'm thinking it must be something to do with linker flags.  `--large-address-aware`?

jeremyd2019 commented:
```
 %{rdynamic: --export-all-symbols}   %{!shared: %{!mdll: --large-address-aware --tsaware}}
```

jeremyd2019 committed:
Misc fixes while testting 32-bit

Don't add --disable-high-entropy-va on 32-bit, it's only supported on
64-bit

For executables, add --large-address-aware for 32-bit, and --tsaware for
all, to match gcc.

Map -rdynamic to --export-all-symbols to match https://cygwin.com/cgit/cygwin-packages/gcc/tree/0004-Cygwin-add-dummy-pthread-tsaware-and-large-address-a.patch

mstorsjo commented on clang/lib/Driver/ToolChain.cpp:
```diff
@@ -684,6 +684,8 @@ static StringRef getArchNameForCompilerRTLib(const ToolChain &TC,
   if (Triple.isOSDarwin())
     return "darwin";
+  if (Triple.isWindowsCygwinEnvironment())
+    return "cygwin";
```
What does the fallback (`getOS()` which defaults to `Triple.getOSName()`) return here, compared with this new value?

And I'm a little unsure how this change fits in with the rest of this change; isn't this change orthogonal to the rest? The rest of the change is about directly calling the linker (while still using the same libgcc etc as otherwise?), while this change is needed for e.g. getting the right directory for compiler-rt builtins and similar? I.e. this change might be ok in itself, but it feels to me like it's a separate change for a separate PR?

tyan0 replied:
My apologies for replying so late.

ToolChain::getOS() returns "windows", since the Triple is "x86_64-pc-windows-cygnus".

After investigating the issue again, I found that the problem was simply caused by llvm/clang being built with `LLVM_ENABLE_PER_TARGET_RUNTIME_DIR=ON` while compiler-rt was built with `LLVM_ENABLE_PER_TARGET_RUNTIME_DIR=OFF`.
Setting both to `LLVM_ENABLE_PER_TARGET_RUNTIME_DIR=ON` works.
Therefore, this fix is not necessary.

However, setting `LLVM_ENABLE_PER_TARGET_RUNTIME_DIR=OFF` for both still does not work. I still investigation that. It seems that clang ignores `LLVM_ENABLE_PER_TARGET_RUNTIME_DIR=OFF`.

tyan0 replied:
I observed following behaviour.

In `ToolChain::getCompilerRT()`, the path `/usr/lib/clang/22/lib/x86_64-pc-windows-cygnus/libclang_rt.builtins.a` is checked first. Then, `/usr/lib/clang/22/lib/windows/libclang_rt.builtins-x86_64.a` is checked.

This algorithm is not affected by `LLVM_ENABLE_PER_TARGET_RUNTIME_DIR` setting in clang.

If `LLVM_ENABLE_PER_TARGET_RUNTIME_DIR` is `ON` in compiler-rt, the first search will succeed. However, if `LLVM_ENABLE_PER_TARGET_RUNTIME_DIR` is `OFF` in compiler-rt, the correct path should not be `/usr/lib/clang/22/lib/windows/libclang_rt.builtins-x86_64.a`, but `/usr/lib/clang/22/lib/cygwin/libclang_rt.builtins-x86_64.a`, therefore, both search will fail. The patch above solves this issue.

As a conclusion, I’ve confirmed that the patch is indeed necessary after all. So, I'll create another PR for that.

tyan0 replied:
Done.
https://github.com/llvm/llvm-project/pull/208925

mstorsjo commented on clang/lib/Driver/ToolChains/Cygwin.cpp:
```diff
@@ -107,3 +110,300 @@ void Cygwin::AddClangSystemIncludeArgs(const ArgList &DriverArgs,
+      CmdArgs.push_back(Args.MakeArgString(ToolChain.GetFilePath("crt0.o")));
+    if (ToolChain.GetRuntimeLibType(Args) == ToolChain::RLT_CompilerRT) {
+      std::string crtbegin =
+          ToolChain.getCompilerRT(Args, "crtbegin", ToolChain::FT_Object);
```
From which toolchain file is this bit copied? We don't have these bits in the mingw toolchain. Has someone tested and used cygwin with `-rtlib=compiler-rt` already, or is this untested theoretical code?

kikairoya replied:
I have local patches to build compiler-rt (@tyan0 might have another work), and succeeded to compile a program with `--rtlib=compiler-rt` (aside from usefulness in current Cygwin toolchain environment).
FWIW, the built binaries are in https://github.com/kikairoya/llvm-cygwin-patches/actions/runs/18583362115 , artifacts named package-*-cygwin-fullpatch.

kikairoya replied:
It appears to attempt to link crtbegin twice when `-rtlib=compiler-rt` and if `clang_rt.crtbegin.o` is available.
The entire "if (RLT_CompilerRT)" block can be removed since compiler-rt doesn't provide crtbegin for Cygwin.

mstorsjo commented on clang/lib/Driver/ToolChains/Cygwin.cpp:
```diff
@@ -107,3 +110,300 @@ void Cygwin::AddClangSystemIncludeArgs(const ArgList &DriverArgs,
+  bool saw_nxcompat = false;
+  for (const char *Arg : CmdArgs) {
+    if (StringRef(Arg) == "-high-entropy-va" ||
+        StringRef(Arg) == "--high-entropy-va" ||
```
I wonder if there's a more clever way to express this, without having to repeat every constant twice, for checking for one or two leading dashes? And the `StringRef(Arg)` wrapping maybe could be stored in a variable instead of remade in each comparison?

jeremyd2019 replied:
thinking about this more, these are last-wins options, so couldn't we just add our intended defaults before the AddLinkerInputs and be done with it?  Or am I forgetting the reason why that wouldn't work?

mstorsjo replied:
I think that should work fine as well.

tyan0 replied:
That looks an excellent idea! I'll try to adopt that. Please wait a while.

tyan0 replied:
Done.

tyan0 committed:
Revert the fix to getOSLibName() to separate it into its own PR

tyan0 committed:
Simplify -high-entropy-va handling

kikairoya commented on clang/lib/Driver/ToolChains/Cygwin.cpp:
https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

> Inconsistent bracing within if/else if/else chains (if one block requires braces, all must)

tyan0 replied:
Fixed.

kikairoya commented on clang/lib/Driver/ToolChains/Cygwin.cpp:
IIRC this predicate has been removed. Have you tested it with the current trunk, after rebasing?

tyan0 replied:
Thanks. I revised the code with reference to MinGW.cpp.

kikairoya commented on clang/lib/Driver/ToolChains/Cygwin.cpp:
This helper has been renamed.

tyan0 replied:
Indeed, although `clang/include/clang/Driver/CommonArgs.h` still has its declaration. Removed.

kikairoya commented on clang/test/Driver/cygwin.cpp:
Does it need to check that `--disable-high-entropy-va` isn't emitted when targeting i686?

tyan0 replied:
Added as well as -disable-nxcompat check.

kikairoya replied:
This check will be done in CHECK-64-ASLR.

kikairoya replied:
You can remove this line if you want to check `--disable-high-entropy-va` in the `CHECK-ASLR`/`CHECK-64-ASLR` blocks, as it is redundant. Otherwise, you can merge those blocks into the `CHECK` block above and `CHECK-64` here, respectively.

tyan0 replied:
Which line do you mean?
L60: // CHECK-64-SAME: "--disable-high-entropy-va"
?

kikairoya replied:
> L60: // CHECK-64-SAME: "--disable-high-entropy-va"

yes.

kikairoya replied:
This is redundant because it's already checked in `CHECK-64-ASLR-DEFAULT`.
```suggestion
```

kikairoya replied:
(assuming https://github.com/llvm/llvm-project/pull/147960/changes#r3743109655 is replying to here)

I'm talking about `// CHECK-...: "--disable-high-entropy-va"`.

```
+// CHECK-64:      "{{.*}}ld{{(\.exe)?}}"
+// CHECK-64-SAME: "-m" "i386pep"
+// CHECK-64-SAME: "{{.*}}{{/|\\\\}}crt0.o"
+// CHECK-64-SAME: "{{.*}}x86_64-pc-msys{{/|\\\\}}{{[0-9.]*}}{{/|\\\\}}crtbegin.o"
+// CHECK-64-SAME: "--disable-high-entropy-va"    <-------------------------------- THIS
+// CHECK-64-SAME: "{{.*}}x86_64-pc-msys{{/|\\\\}}{{[0-9.]*}}{{/|\\\\}}crtend.o"
```

```
+// CHECK-64-ASLR-DEFAULT:      "{{.*}}ld{{(\.exe)?}}"
+// CHECK-64-ASLR-DEFAULT-SAME: "--disable-high-entropy-va" <----- AND THIS
+// CHECK-64-ASLR-DEFAULT-SAME: "--disable-nxcompat"
```

Why is `--disable-high-entropy-va` needed in CHECK-64, especially since `--disable-nxcompat`, which is paired in `CHECK-64-ASLR-DEFAULT,` isn't checked?

tyan0 replied:
Ah, indeed. I removed the one in CHECK-64. Thanks!

kikairoya commented on clang/test/Driver/cygwin.cpp:
I think this check is redundant now.

tyan0 replied:
Removed. Thanks.

kikairoya commented on clang/test/Driver/cygwin.cpp:
ditto.

tyan0 replied:
Dotto.

kikairoya commented on clang/lib/Driver/ToolChains/Cygwin.cpp:
Since `OutputFile` isn't used after this block, it can be simplified like this:
```c++
if (StringRef OutputFile = ...; !has_extension(OutputFile))
  CmdArgs.emplace_back(...);
else
  CmdArgs.push_back(OutputFile);
```

tyan0 replied:
Done.

tyan0 committed:
Updated to work with the latest main branch

tyan0 committed:
Update llvm-lit test for Cygwin to reflect the driver changes

tyan0 committed:
Simplify `OutputFile` handling in clang cygwin driver a bit

kikairoya commented on clang/test/Driver/cygwin.cpp:
```diff
@@ -54,24 +55,69 @@
+// RUN:   2>&1 | FileCheck --check-prefix=CHECK-ASLR-DEFAULT %s
+// CHECK-ASLR-DEFAULT:      "{{.*}}ld{{(\.exe)?}}"
+// CHECK-ASLR-DEFAULT-NOT:  "--disable-high-entropy-va"
+// CHECK-ASLR-DEFAULT-SAME: "--disable-nxcompat"
```
This check fails to catch when the order of high-entropy-va and nxcompat is swapped.
We need to repeat -NOT after all positive checks, repeat RUN lines to split -NOT checks and positive checks, or use `FileCheck --implicit-check-not`.

kikairoya commented on clang/test/Driver/cygwin.cpp:
```diff
@@ -54,24 +55,69 @@
+// RUN: %clang -### %s --target=x86_64-pc-cygwin --sysroot=%S/Inputs/basic_cygwin_tree \
+// RUN:   2>&1 | FileCheck --check-prefix=CHECK-64-ASLR-DEFAULT %s
+// CHECK-64-ASLR-DEFAULT:      "{{.*}}ld{{(\.exe)?}}"
+// CHECK-64-ASLR-DEFAULT-SAME: "--disable-high-entropy-va"
```
```suggestion
// CHECK-64-ASLR-DEFAULT-SAME: "--disable-high-entropy-va"
// CHECK-64-ASLR-DEFAULT-SAME: "--disable-nxcompat"
```

kikairoya commented on clang/lib/Driver/ToolChains/Cygwin.cpp:
```diff
@@ -107,3 +110,284 @@ void Cygwin::AddClangSystemIncludeArgs(const ArgList &DriverArgs,
+    CmdArgs.push_back("console");
+  }
+
+  if (ToolChain.getTriple().isArch32Bit()) {
+    CmdArgs.push_back("--wrap=_Znwj");
+    CmdArgs.push_back("--wrap=_Znaj");
+    CmdArgs.push_back("--wrap=_ZdlPv");
+    CmdArgs.push_back("--wrap=_ZdaPv");
+    CmdArgs.push_back("--wrap=_ZnwjRKSt9nothrow_t");
+    CmdArgs.push_back("--wrap=_ZnajRKSt9nothrow_t");
+    CmdArgs.push_back("--wrap=_ZdlPvRKSt9nothrow_t");
+    CmdArgs.push_back("--wrap=_ZdaPvRKSt9nothrow_t");
+  } else {
+    CmdArgs.push_back("--wrap=_Znwm");
+    CmdArgs.push_back("--wrap=_Znam");
+    CmdArgs.push_back("--wrap=_ZdlPv");
+    CmdArgs.push_back("--wrap=_ZdaPv");
+    CmdArgs.push_back("--wrap=_ZnwmRKSt9nothrow_t");
+    CmdArgs.push_back("--wrap=_ZnamRKSt9nothrow_t");
+    CmdArgs.push_back("--wrap=_ZdlPvRKSt9nothrow_t");
+    CmdArgs.push_back("--wrap=_ZdaPvRKSt9nothrow_t");
+  }
```
A test for this block might be needed.

tyan0 replied:
Thanks for the advice. Done.

kikairoya reviewed (commented):
I'm sorry for commenting bit by bit.

tyan0 committed:
Update clang/test/Driver/cygwin.cpp

Co-authored-by: Tomohiro Kashiwada <kikairoya@gmail.com>

kikairoya commented on clang/test/Driver/cygwin.cpp:
```diff
@@ -54,24 +55,89 @@
+// RUN: %clang -### %s --target=i686-pc-cygwin --sysroot=%S/Inputs/basic_cygwin_tree \
+// RUN:   2>&1 | FileCheck --check-prefix=CHECK-ASLR-DEFAULT %s
+// CHECK-ASLR-DEFAULT:      "{{.*}}ld{{(\.exe)?}}"
+// CHECK-ASLR-DEFAULT-NOT:  "--disable-high-entropy-va"
```
Why was the `--disable-nxcompat` check removed?

tyan0 replied:
Umm, it seems that I've misread your comment. I'll revive `--disable-nxcompat` check.

kikairoya commented on clang/test/Driver/cygwin.cpp:
```diff
@@ -54,24 +55,89 @@
+// CHECK-64-ASLR-DEFAULT-SAME: "--wrap=_ZnwmRKSt9nothrow_t"
+// CHECK-64-ASLR-DEFAULT-SAME: "--wrap=_ZnamRKSt9nothrow_t"
+// CHECK-64-ASLR-DEFAULT-SAME: "--wrap=_ZdlPvRKSt9nothrow_t"
+// CHECK-64-ASLR-DEFAULT-SAME: "--wrap=_ZdaPvRKSt9nothrow_t"
```
These checks don't seem to relate to "ASLR".

tyan0 replied:
Right. It seems that I misunderstood your comment: "This check will be done in CHECK-64-ASLR.". I'll add another check for `--wrap=xxx` things.

kikairoya reviewed (commented):
Are there no tests using crt0.o and crtend.o?

tyan0 committed:
Update clang/test/Driver/cygwin.cpp again

- Add crt0, crtbegin0, and crtendo check
- Add dummy crt0 files for linux-cross environment
- Revive `--disable-nxcompat` check
- Separate the `wrap=xxx` checks from CHECK-64-ASLR-DEFAULT into
  CHECK-64-WRAP

tyan0 committed:
Change the location of crt0.o for linux cross cygwin environment

Previously, they were:
basic_cross_cygwin_tree/usr/lib/gcc/i686-pc-msys/10/crt0.o
basic_cross_cygwin_tree/usr/lib/gcc/x86_64-pc-cygwin/10/crt0.o

These have benn moved to:
basic_cross_cygwin_tree/usr/i686-pc-msys/lib/crt0.o
basic_cross_cygwin_tree/usr/x86_64-pc-cygwin/lib/crt0.o

kikairoya commented on clang/lib/Driver/ToolChains/Cygwin.cpp:
```diff
@@ -107,3 +110,284 @@ void Cygwin::AddClangSystemIncludeArgs(const ArgList &DriverArgs,
+          CmdArgs.push_back(Args.MakeArgString(P));
+        }
+      }
+      CmdArgs.push_back(Args.MakeArgString(ToolChain.GetFilePath("crtend.o")));
```
The same as crtbegin. compiler-rt doesn't provide crtend.

tyan0 replied:
> This is redundant because it's already checked in CHECK-64-ASLR-DEFAULT.

It looks like some quotation is broken.
What does _"This"_ mean?

tyan0 replied:
> It appears to attempt to link crtbegin twice when -rtlib=compiler-rt and if clang_rt.crtbegin.o is available.
The entire "if (RLT_CompilerRT)" block can be removed since compiler-rt doesn't provide crtbegin for Cygwin.
> The same as crtbegin. compiler-rt doesn't provide crtend.

Thanks. I confirmed the driver works without crtbegin/crtend.
Revised the patch.

kikairoya reviewed (commented):
Sorry for the delay.

Now that the other PR has been merged, we can add some tests for `-rtlib=compiler-rt`.

tyan0 committed:
Do not link crtbegin and crtend when compiler-rt is used

tyan0 committed:
Update clang/test/Driver/cygwin.cpp

remove redundant check for --disable-high-entropy-va for 64bit
environment.

tyan0 committed:
Update clang/test/Driver/cygwin.cpp

Add check for -rtlib=compiler-rt option

mstorsjo commented on clang/lib/Driver/ToolChains/Cygwin.cpp:
```diff
@@ -107,3 +110,265 @@ void Cygwin::AddClangSystemIncludeArgs(const ArgList &DriverArgs,
+  if (!Args.hasArg(options::OPT_shared) && !Args.hasArg(options::OPT_mdll)) {
+    if (ToolChain.getTriple().isArch32Bit())
+      CmdArgs.push_back("--large-address-aware");
+    CmdArgs.push_back("--tsaware");
```
This logic for `--large-address-aware` and `--tsaware` seems to be different from what we have in e.g. the mingw target. Could you add a comment here to explain why these are added, and why they aren't added if linking a DLL? (I presume the answer is quite cygwin specific, so it would be useful info for people unfamiliar with it looking at the code.)

tyan0 replied:
Add commets explaining why these options are added for the consitions here. Thanks.

mstorsjo commented on clang/lib/Driver/ToolChains/Cygwin.cpp:
```diff
@@ -107,3 +110,265 @@ void Cygwin::AddClangSystemIncludeArgs(const ArgList &DriverArgs,
+  if (Args.hasArg(options::OPT_Z_Xlinker__no_demangle))
+    CmdArgs.push_back("--no-demangle");
+
+  // Default. could be override by `last-wins`.
```
This comment seems to have slightly broken structure. Should it be "Could be overridden"? The "last-wins" part also seems a bit unclear in this form. What about "Default. Can be overridden by the user by explicitly passing an option after this one." or something like that?

tyan0 replied:
Revised the comment as you suggested. Thanks.

mstorsjo reviewed (commented):
Looking mostly good here, thanks! Just two minor requests about comments.

tyan0 committed:
Update clang/test/Driver/cygwin.cpp source comment

- Add explanations for why the option --large-address-aware and
  --tsaware options are required on Cygwin.
- Revise the comment on defaulting to --disable-high-entropy-va.

mstorsjo reviewed (approved):
LGTM now, thanks!
